TechforgoodCAST/fusebox-insights
# Support messages have rules

Few bits to this PR so I'll try to give a bit of detail as to what's going on and provide steps to test it.

- support messages now have fields `rule_object_type`,  `rule_event_type` and `rule_occurrences`
`rule_object_type` = the type of object that an event relates to e.g. Unknown
`rule_event_type` = type of event e.g. create
`rule_occurrences` = number of events that need to happen for the support message to be resolved

- a callback now fires when a new `Event` is created. This callback searches for all events of the same type for the same project to count how many times this type of event has happened. It then checks for incomplete support messages for the corresponding project that would be resolved given the events that have been logged. These are then updated to `status=complete`.

- An interesting edge case for this is if a support message is created with a rule `create 2 Unknowns`, two unknowns are then created, and the support message is marked as `complete`, if then a second support message is created with a rule with the same object and event type (`create 2 Unknowns` again for example). As soon as a new event is triggered for the creation of a new Unknown, the second support message would be resolved, as there would be 3 event logs for the creation of unknowns in the DB. To fix this, an `accounted_for` field has been added to the `event` model. The callback only looks for events where `accounted_for=false` to check against support message rules. The relevant events are marked as `accounted_for=true` if they are used to resolve a support message. This means old events that have closed previous messages will not be counted in future callbacks.

Example To Test:

- create user with `is_staff=true`
- login and create project
- create a new support message with a rule `create 2 unknowns` (this should then display in the projects index view)
- create two new unknowns for the project (added a dropdown for now to quickly add association)
- support message should now be `Complete`
- create a second support message with a similar rule (e.g. `create 2 unknowns`) and follow the same process to see that existing event logs won't be taken into account for the new support message
